### PR TITLE
[ML] Multinomial logistic regression

### LINKED ISSUE #982: [ML] Support multinomial logistic regression in data frame analysis
This issue tracks introducing multinomial logistic regression. The main complication of this is that the loss function has multiple parameters and the forest needs to predict all of them. This breaks the assumption that we have a single prediction, gradient and curvature. The task therefore breaks down into some preliminary tasks independent of actually implementing the new loss function.

- [x] Extend `CLoss` and `CArgMinLoss` to support multivalued predictions, gradients and a Hessian matrix
- [x] Rework caching to store multivalued predictions, gradients and the Hessian. This affects how we index values in the data frame, the memory calculation, etc
- [x] Rework the code to compute aggregate derivatives for candidate splits
- [x] Rework the code to choose the best candidate split
- [x] Extend `CBoostedTreeNode` to support multi-valued leaves
- [x] Extend the inference model definition to support multi-valued leaves
- [x] Extend feature importance to support multi-valued leaves
- [x] Write the new loss function
- [x] Wire in to classification api

### PR DESCRIPTION
**Pull request by tveasey**
This change implements the loss function for multinomial logistic regression.

Note I've factored out the loss function related unit tests into their own suite. I also needed to make various changes to our online k-means implementation to support `CDenseVector` which requires that the vector dimension is passed to the constructor.

### CODE REVIEW

File: lib/maths/CBoostedTreeLoss.cc [hunk 1]
```diff
@@ -205,6 +223,212 @@ CArgMinLogisticImpl::TDoubleVector CArgMinLogisticImpl::value() const {
...
+                                             double actual,
+                                             double weight) {
+
+    using TMinAccumulator = CBasicStatistics::SMin<std::pair<double, std::size_t>>::TAccumulator;
+
+    switch (m_CurrentPass) {
+    case 0: {
+        // We have a member variable to avoid allocating a tempory each time.
+        m_DoublePrediction = prediction;
+        m_PredictionSketch.add(m_DoublePrediction, weight);
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeLoss.cc]: The basic idea is as follows. We want to find the `weight = argmin_w { sum_{i in I}{ [-log([softmax(prediction_i + w)]_{actual_i}]) } + lambda * w'w }`. Here, `I` is the set of all training examples in the leaf and `actual_i` is the index of the i'th example actual category. Rather than working with this function directly we summarise it by the set `{(x_j, c_j)}` where `x_j` are some points in prediction space and `c_j` are the counts of the nearest predictions in the `I` to each `x_j`. We use a k-means of the predictions in `I` to choose `x_j` this is calculated sequentially (to accommodate the case we're using disk storage). This is what `m_PredictionSketch` is doing. I'll add some class documentation to explain the strategy.

File: lib/maths/CBoostedTreeLoss.cc [hunk 2]
```diff
@@ -205,6 +223,212 @@ CArgMinLogisticImpl::TDoubleVector CArgMinLogisticImpl::value() const {
...
+    for (std::size_t i = 0; i < NUMBER_RESTARTS; ++i) {
+        for (int j = 0; j < x0.size(); ++j) {
+            double alpha{CSampling::uniformSample(m_Rng, 0.0, 1.0)};
+            x0(j) = weightBoundingBox[0](j) +
+                    alpha * (weightBoundingBox[1](j) - weightBoundingBox[0](j));
+        }
+        LOG_TRACE(<< "x0 = " << x0.transpose());
+
+        double loss;
+        CLbfgs<TDoubleVector> lgbfs{5};
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CBoostedTreeLoss.cc]: This is the rank of the Hessian approximation. Generally, people recommend nothing too small (less than 3), but you quickly get diminishing returns. For example, 5 is the default for the R's optim package for this parameter. We can experiment a bit with this, but [this](https://github.com/elastic/ml-cpp/pull/1037/files#diff-db63fc7b8848a848d40ea4a725899f87R503) randomised test suggests the choice isn't too bad.

File: include/maths/CBoostedTreeLoss.h [hunk 3]
```diff
@@ -95,15 +99,55 @@ class MATHS_EXPORT CArgMinLogisticImpl final : public CArgMinLossImpl {
...
     TMinMaxAccumulator m_PredictionMinMax;
-    TDoubleVector2x1 m_CategoryCounts;
-    TDoubleVector2x1Vec m_BucketCategoryCounts;
+    TDoubleVector2x1 m_ClassCounts;
+    TDoubleVector2x1Vec m_BucketsClassCounts;
+};
+
+//! \brief Finds the value to add to a set of predicted multinomial logit which
+//! minimises regularised cross entropy loss w.r.t. the actual classes.
+class MATHS_EXPORT CArgMinMultinomialLogisticLossImpl final : public CArgMinLossImpl {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply include/maths/CBoostedTreeLoss.h]: +1. I added an implementation note in [d59a981](https://github.com/elastic/ml-cpp/pull/1037/commits/d59a981317dd0b2f43bc8dc7a43d3a0e19522de4).

**Comment by tveasey** [comment]
Thanks for the review @valeriy42. I've been through your comments and added an explanation of the top level strategy. Regarding,
> I haven't seen you normalizing the output probability vector. Did I miss it or you don't want to normalize?

I don't actually normalise values of weights (although with "shrinkage" regularisation they shouldn't get too big). Also, I haven't actually wired this in to `api::CDataFrameTrainBoostedTreeClassifierRunner` yet. This PR is already getting large, so I think I will postpone that for the next one. I'll also finish up the last bit of loss function testing and add a multi-class unit test. I'm therefore also removing WIP.

File: lib/maths/unittest/CBoostedTreeLossTest.cc [hunk 4]
```diff
@@ -476,6 +501,98 @@ BOOST_AUTO_TEST_CASE(testMultinomialLogisticMinimizerEdgeCases) {
 }
 
 BOOST_AUTO_TEST_CASE(testMultinomialLogisticMinimizerRandom) {
+
+    // Test that we a good approximation of the additive term for the log-odds
```
Comment (valeriy42) [inline lib/maths/unittest/CBoostedTreeLossTest.cc]:
```suggestion
    // Test that we have a good approximation of the additive term for the log-odds
```
